### ISSUE TITLE
Assertion truncation: properly handle long lines

### ISSUE THREAD

**Issue by mattduck** [pytest-dev/pytest#1954]
(I'm running on the features branch).

Hey @flub, thanks for running the sprint the other day. I had a thought earlier: long lines can break the assertion truncation code. This is because we're checking the length of all lines [here](https://github.com/pytest-dev/pytest/blob/f279bd2bd5871b05e95e49e79fd9dff54e16e923/_pytest/assertion/__init__.py#L134), but after that we assume that there are [at least 10 lines in the explanation list](https://github.com/pytest-dev/pytest/blob/f279bd2bd5871b05e95e49e79fd9dff54e16e923/_pytest/assertion/__init__.py#L140).

Previously the only consequence of this is that if you had a list with a small number of long lines, you could trigger a nonsensical message:

<img width="547" alt="screen shot 2016-09-22 at 00 33 50" src="https://cloud.githubusercontent.com/assets/1607892/18732595/4b830ef4-805c-11e6-9088-766e6194a74c.png">

However, since 999e7c65417f1e97fc89bf66e0da4c5cd84442ec **it causes an IndexError**.

Example:

```
def test_github():
    a = 'a' * 1000
    b = 'b' * 1000
    assert a == b
```

Produces:

```
...
                # Truncate lines if required
                if (sum(len(p) for p in new_expl[1:]) > 80*8 and
                        item.config.option.verbose < 2 and
                        not _running_on_ci()):
                    show_max = 10
                    truncated_count = len(new_expl) - show_max
>                   new_expl[show_max - 1] += " ..."
E                   IndexError: list index out of range
```

---

I think we should add test cases for this situation to ensure that long lines are properly supported. I also think the truncation logic will be easier to read and maintain if we extract it into its own function.

I've [started to work on it](https://github.com/mattduck/pytest/commit/72acda12f607be4874ac2caea0f0c0a605666c5d). Still WIP but I'm happy to continue with it.

**Comment by mattduck** [comment]
Also - I think it makes sense to truncate individual lines if they're too long, as they can produce a lot of output even if limited to 10 lines. So without the `-vv` flag, behaviour could be to both (1) truncate at 10 lines, and (2) ensure that only the first n characters are displayed in each line (eg. more than 500 chars seems excessive).

**Comment by flub** [comment]
Splitting the truncation off to it's own function would be nice indeed.  The existing truncation has always been very rough, I'm kind of surprised no one's complained about it yet ;-)

How about collapsing both metrics in one though?  I.e the truncation assumes that you only want to show 8 lines vertically, so just compute that.  If all of those 8 lines are used up by the first line, then only print the first line.  If the first line is longer, then truncate the first line etc.  So just truncate in the text where printing would fill more the 8 lines.

**Comment by mattduck** [comment]
Yeah, I think this is a sensible approach. Will take a look at it.

**Comment by RonnyPfannschmidt** [comment]
i think given the complexity of the code involved we should absolutely pull it into a own function and unittest it

**Comment by mattduck** [comment]
Hey guys, sorry for lack of activity on this - hectic couple of weeks. Will aim to pick it up this week.

**Comment by nicoddemus** [comment]
No worries, thanks for following up

**Comment by nicoddemus** [comment]
Fixed in #1995
